Update etcd performance information and guidance

jeremyeder opened the pull request:
This PR is ready for wide review.

jeremyeder commented:
@smarterclayton @liggitt @enj @mfojtik @derekwaynecarr @sdodson @ingvagabund @deads2k @mffiedler

smarterclayton commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,94 @@ include writing a large amount of log data.
 etcd is a distributed key-value store that {product-title} uses for
 configuration.
 
+Version 3 of etcd brings with it important scalability and performance improvements that reduce CPU, memory, network and disk requirements for any size cluster.  OpenShift 3.5 introduced etcd version 3.1.
```
This sentence should clarify and establish terminology for the rest:

1. release version of etcd (3.x)
2. storage schema (v2 vs v3)

and clearly use it.  This is very confusing for people.  I would recommend saying "etcd 3.x" and "etcd storage v2" and "etcd storage v3".

sdodson replied:
`Version 3 of etcd introduces a new storage version which brings with it important scalability and performance improvements that reduce CPU, memory, network and disk requirements for any size cluster.`

Trying to differentiate between package version and storage api.

I'd get rid of the last sentence, while they may have been upgraded to etcd 3.1, we don't use v3 storage there so i'd just skip it.

jeremyeder replied:
Last sentence dropped.  And I've added a table so it makes the sentence redundant anyway.

jeremyeder replied:
@stevekuznetsov good catch, I have updated the graph to start the Y-axis at 0.

jeremyeder replied:
Sure.  Updated to clarify.  See if you like it.

sdodson commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,94 @@ include writing a large amount of log data.
 
+Version 3 of etcd brings with it important scalability and performance improvements that reduce CPU, memory, network and disk requirements for any size cluster.  OpenShift 3.5 introduced etcd version 3.1.
+
+Version 3 of etcd implements a backwards compatible storage API that facilitates a two-step migration of the on-disk etcd database.  For migration purposes, the storage mode used by etcd 3.1 in OpenShift 3.5 remained in v2 mode.  As of OpenShift 3.6, new installs will use storage mode v3.  Upgrades from previous versions of OpenShift will NOT automatically migrate data from v2 to v3.  This must b ...
```
I'd say `For migration purposes, the storage mode used by OpenShift 3.5 was v2 mode even if etcd 3.x packages were used.` because we don't actually do anything to bound the etcd package version so they may have 3.0, 3.1, or soon 3.2.

sdodson replied:
Or if we get rid of the last sentence in the previous paragraph we could forego this sentence too.

jeremyeder replied:
@sdodson @smarterclayton I added a table.  I think it's easier to convey rather than a long-winded paragraph.

stevekuznetsov commented:
`scaling_performance/images/average-read-write-iops.png` is misleading -- should start at 0 to correctly show that there are half as many IOPS in v3, instead of starting at ~300 and implying there are basically none at all.

smarterclayton commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,94 @@ include writing a large amount of log data.
+
+Version 3 of etcd implements a backwards compatible storage API that facilitates a two-step migration of the on-disk etcd database.  For migration purposes, the storage mode used by etcd 3.1 in OpenShift 3.5 remained in v2 mode.  As of OpenShift 3.6, new installs will use storage mode v3.  Upgrades from previous versions of OpenShift will NOT automatically migrate data from v2 to v3.  This must b ...
+
+In addition to changing the storage mode for new installs to v3, OpenShift 3.6 also begins enforcing "quorum reads" for all OpenShift types.  This is done to ensure that queries against etcd do not return stale data.  In single-node etcd clusters, stale data is not a concern.  In highly available etcd deployments typically found in production clusters, quorum reads ensure valid query results.  Qu ...
```
Last sentence I would change to:

> A quorum read is "linearizable" in database terms - every client sees the latest updated state of the cluster, and all clients see the same sequence of reads and writes.

jeremyeder replied:
Fixed.

smarterclayton commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,94 @@ include writing a large amount of log data.
+
+In addition to changing the storage mode for new installs to v3, OpenShift 3.6 also begins enforcing "quorum reads" for all OpenShift types.  This is done to ensure that queries against etcd do not return stale data.  In single-node etcd clusters, stale data is not a concern.  In highly available etcd deployments typically found in production clusters, quorum reads ensure valid query results.  Qu ...
+
+It is important to note that OpenShift uses etcd for storing additional information beyone what Kubernetes itself requires.  For example, OpenShift stores information about images, builds, RBAC, secrets (amongst others) in etcd as is required by features that OpenShift adds on top of Kubernetes.  Ultimately what this means is that guidance around performance and sizing for etcd hosts will differ  ...
```
RBAC and secrets are in kube, which makes this confusing.

jeremyeder replied:
removed

smarterclayton commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,94 @@ include writing a large amount of log data.
+
+Performance improvements have been quantified using a 300 node OpenShift 3.6 cluster using the https://github.com/openshift/svt/tree/master/openshift_scalability[cluster-loader] utility.
+
+Comparing etcd 3.1 (storage mode v2) versus etcd 3.1 (storage mode v3), clear improvements are identified.  Based on extensive test results, Red Hat strongly recommends migrating existing OpenShift clusters to etcd storage mode v3, particularly for larger clusters, or in scenarios where SSD storage is not available.
```
The challenge here is that this isn't a choice.  In 3.6 you *must* upgrade to v3 storage mode.  Period.  We should describe this in terms of *why* you are being forced to upgrade, but not leave open the door for people to not think they have to update.

jeremyeder replied:
Maybe I misunderstood something but was told that the migration wasn't automatic when you upgrade from 3.5->3.6.   So is it the case that after upgrading you have to do a manual run of the migration playbooks before the cluster is usable?  This morning I asked @ingvagabund for docs on the migration approach..that'd clear everything up for me.

jeremyeder replied:
I've updated the docs according to our discussions.  I put in an "IMPORTANT" box stating that future versions of OpenShift will require an etcd schema update.

ashishkamra commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,94 @@ include writing a large amount of log data.
+
+Version 3 of etcd implements a backwards compatible storage API that facilitates a two-step migration of the on-disk etcd database.  For migration purposes, the storage mode used by etcd 3.1 in OpenShift 3.5 remained in v2 mode.  As of OpenShift 3.6, new installs will use storage mode v3.  Upgrades from previous versions of OpenShift will NOT automatically migrate data from v2 to v3.  This must b ...
+
+In addition to changing the storage mode for new installs to v3, OpenShift 3.6 also begins enforcing "quorum reads" for all OpenShift types.  This is done to ensure that queries against etcd do not return stale data.  In single-node etcd clusters, stale data is not a concern.  In highly available etcd deployments typically found in production clusters, quorum reads ensure valid query results.  Qu ...
```
As per [1], "While etcd 3.0 processes linearized reads through direct raft proposals, chewing precious disk bandwidth and incurring the corresponding latency penalty, linearized reads in 3.1 issue idempotent, fsync-free linearized raft index requests. These index requests are also batched; concurrent linearized reads coalesce into a single index request. The upshot is linearized reads have lower latency and better throughput."

So it is probably incorrect to state that 'Quorum reads force an fsync to guarantee all nodes are in sync' as that does not seem to be the case in etcd 3.1. May be a better idea to refer to [1] for the details on how quorum reads are handled in etcd 3.1?

[1] https://coreos.com/blog/etcd-3-1-announcement.html

jeremyeder replied:
Thank you, I have fixed it accordingly.

soltysh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,103 @@ include writing a large amount of log data.
+|3.6 |3.x | v3
+|======================
+
+<<<<<<< HEAD
```
This looks like a leftover from rebase.

jeremyeder replied:
Woops, good catch, I've removed them.

sdodson commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,101 @@ include writing a large amount of log data.
+|======================
+|OpenShift Version |etcd version|storage schema version
+|3.4 and earlier |2.x | v2
+|3.5 |3.x | v2
```
Lets change this to `3.3 and earlier` and `3.4 and 3.5`. The 3.4 upgrade playbooks will upgrade you to at least etcd 3.0.

I don't think it's worth making it even more confusing, but we don't do anything to bound the version of etcd installed so even a clean install of 3.1 today would get etcd 3.1 or whatever the latest is at install time.

jeremyeder replied:
Updated according to what I think you want.

sdodson replied:
looks good to me

ahardin-rh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,106 @@ include writing a large amount of log data.
+|3.6 |3.x | v3 (new installs)
+|======================
+
+etcd 3.x brings with it important scalability and performance improvements that reduce CPU, memory, network and disk requirements for any size cluster.  etcd 3.x also implements a backwards compatible storage API that facilitates a two-step migration of the on-disk etcd database.  For migration purposes, the storage mode used by etcd 3.x in OpenShift 3.5 remained in v2 mode.  As of OpenShift 3.6, ...
```
s/brings with it/introduces
network, and disk requirements
s/installs/installations
s/OpenShift/{product-title}
{product-title} 3.5
{product-title} 3.6
You must use the supplied playbooks and follow the documented process to migrate the data.

        s/NOT/ _not_

ahardin-rh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,106 @@ include writing a large amount of log data.
+etcd 3.x brings with it important scalability and performance improvements that reduce CPU, memory, network and disk requirements for any size cluster.  etcd 3.x also implements a backwards compatible storage API that facilitates a two-step migration of the on-disk etcd database.  For migration purposes, the storage mode used by etcd 3.x in OpenShift 3.5 remained in v2 mode.  As of OpenShift 3.6, ...
+
+Version 3 of etcd implements a backwards compatible storage API that facilitates a two-step migration of the on-disk etcd database.  For migration purposes, the storage mode used by etcd 3.x in OpenShift 3.5 remained in v2 mode.  As of OpenShift 3.6, new installs will use storage mode v3.  In order to provide customers time to prepare for migrating the etcd schema from v2 to v3 (and associated do ...
+
```
s/installs/installations
{product-title} 3.5
{product-title} 3.6
results,

ahardin-rh replied:
Use only single spaces between sentences

ahardin-rh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,106 @@ include writing a large amount of log data.
+
+[IMPORTANT]
+====
+etcd schema migration will be required by future OpenShift upgrades.
```
s/OpenShift/{product-title}

ahardin-rh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,106 @@ include writing a large amount of log data.
+etcd schema migration will be required by future OpenShift upgrades.
+====
+
+In addition to changing the storage mode for new installs to v3, OpenShift 3.6 also begins enforcing "quorum reads" for all OpenShift types.  This is done to ensure that queries against etcd do not return stale data.  In single-node etcd clusters, stale data is not a concern.  In highly available etcd deployments typically found in production clusters, quorum reads ensure valid query results.  A  ...
```
s/installs/installations
s/OpenShift/{product-title}

    s/"quorum reads"/_quorum reads_  (only the first instance of usage)
    s/"linearizable"/_inearizable_   (only the first instance of usage)

s/-/--
link:https://coreos.com/blog/etcd-3-1-announcement.html[announcement]

ahardin-rh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,106 @@ include writing a large amount of log data.
+
+In addition to changing the storage mode for new installs to v3, OpenShift 3.6 also begins enforcing "quorum reads" for all OpenShift types.  This is done to ensure that queries against etcd do not return stale data.  In single-node etcd clusters, stale data is not a concern.  In highly available etcd deployments typically found in production clusters, quorum reads ensure valid query results.  A  ...
+
+It is important to note that OpenShift uses etcd for storing additional information beyond what Kubernetes itself requires.  For example, OpenShift stores information about images, builds (amongst others) in etcd as is required by features that OpenShift adds on top of Kubernetes.  Ultimately what this means is that guidance around performance and sizing for etcd hosts will differ from Kubernetes ...
```
s/OpenShift/{product-title}
s/builds (amongst others)/ builds, and other components 
etcd, as is required
Ultimately, this means that

ahardin-rh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,106 @@ include writing a large amount of log data.
+
+It is important to note that OpenShift uses etcd for storing additional information beyond what Kubernetes itself requires.  For example, OpenShift stores information about images, builds (amongst others) in etcd as is required by features that OpenShift adds on top of Kubernetes.  Ultimately what this means is that guidance around performance and sizing for etcd hosts will differ from Kubernetes ...
+
+Performance improvements have been quantified using a 300 node OpenShift 3.6 cluster using the https://github.com/openshift/svt/tree/master/openshift_scalability[cluster-loader] utility.
```
s/have been/were
300-node
s/OpenShift/{product-title}
link:https://github.com/openshift/svt/tree/master/openshift_scalability[cluster-loader]

ahardin-rh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,106 @@ include writing a large amount of log data.
+image::average-read-write-iops.png["Read+Write IOPS"]
+endif::[]
+
+CPU utilization by both the apiserver (master) and etcd processes is reduced:
```
s/apiserver/ API server

ahardin-rh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,106 @@ include writing a large amount of log data.
+image::cpu-usage.png["CPU Usage"]
+endif::[]
+
+Memory utilization by both the apiserver (master) and etcd processes is also reduced:
```
s/apiserver/ API server

ahardin-rh commented on scaling_performance/host_practices.adoc:
```diff
@@ -57,42 +57,106 @@ include writing a large amount of log data.
 [NOTE]
 ====
-etcd hosts are typically memory intensive.
+etcd processes are typically memory intensive.  master/apiserver processes are CPU intensive.  This makes them a reasonable co-location pair within a single machine/VM.  Optimize communication between etcd and master hosts either by co-locating them on the same host, or providing a dedicated network.
```
Master / API server
single machine or virtual machine (VM)

ahardin-rh commented:
@jeremyeder Just minor nits from me. Once those changes are made and everything is squashed into 1 commit, we should be able to merge.

jeremyeder commented:
@ahardin-rh @smarterclayton @sdodson @soltysh @ashishkamra @stevekuznetsov addressed all comments and squashed.  Thank you for the review and attention to detail.

ahardin-rh commented:
[rev_history]
|xref:../scaling_performance/host_practices.adoc#scaling-performance-capacity-host-practices[Recommended Host Practices]
|Updated etcd performance guidance.
%
